File: baselib/src/main/java/com/lishuaihua/baselib/util/ScreenShotUtils.java
package com.lishuaihua.baselib.util;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Point;
import android.graphics.Rect;
import android.os.Build;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.webkit.WebView;
import android.widget.ListView;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

import androidx.core.widget.NestedScrollView;


public class ScreenShotUtils {
    private static String TAG = "ScreenShotUtils";


    // 获取指定Activity的截屏，保存到png文件
    public static Bitmap takeScreenShot(Activity activity,int navigationBarHeight) {
        // View是你需要截图的View
        View view = activity.getWindow().getDecorView();
        view.setDrawingCacheEnabled(true);
        view.buildDrawingCache();
        Bitmap b1 = view.getDrawingCache();

        // 获取状态栏高度
        Rect frame = new Rect();
        activity.getWindow().getDecorView().getWindowVisibleDisplayFrame(frame);
        int statusBarHeight = frame.top;
        System.out.println(statusBarHeight);

        // 获取屏幕长和高
        int width = activity.getWindowManager().getDefaultDisplay().getWidth();
        int height = activity.getWindowManager().getDefaultDisplay()
                .getHeight();
        // 去掉标题栏
        // Bitmap b = Bitmap.createBitmap(b1, 0, 25, 320, 455);
        Bitmap b = Bitmap.createBitmap(b1, 0, statusBarHeight+navigationBarHeight, width, height
                - statusBarHeight-navigationBarHeight);
        view.destroyDrawingCache();
        return b;
    }

    // 保存到sdcard
    public static void savePic(Bitmap b, String strFileName) {
        FileOutputStream fos = null;
        Log.d("TAG", "savePic() returned: "  );
        try {
            fos = new FileOutputStream(strFileName);
            if (null != fos) {
                b.compress(Bitmap.CompressFormat.PNG, 90, fos);
                Log.d("TAG", "savePic() returned:    " + b.getHeight());
                fos.flush();
                fos.close();
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // 截取当前屏幕
    public static void shootScreenView(Activity a, String picpath) {
        ScreenShotUtils.savePic(ScreenShotUtils.takeScreenShot(a,0), picpath);
    }

    /**
     * 保存bitmap到SD卡
     * @param path 保存的名字
     * @param mBitmap 图片对像
     * return 生成压缩图片后的图片路径
     */
    public static boolean saveMyBitmap(Bitmap mBitmap, String path) {
        File f = new File(path);

        try {
            f.createNewFile();
        } catch (IOException e) {
            System.out.println("在保存图片时出错：" + e.toString());
            return false;

        }
        FileOutputStream fOut = null;
        try {
            fOut = new FileOutputStream(f);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            return false;

        }
        try {
            mBitmap.compress(Bitmap.CompressFormat.PNG, 100, fOut);
        } catch (Exception e) {
            return false;

        }
        try {
            fOut.flush();
        } catch (IOException e) {
            e.printStackTrace();
            return false;

        }
        try {
            fOut.close();
        } catch (IOException e) {
            e.printStackTrace();
            return false;

        }

        return true;
    }



    // 程序入口 截取ScrollView
    public static boolean shotNestedScrollView(NestedScrollView scrollView, String picpath ) {
        return ScreenShotUtils.saveMyBitmap(getNestedScrollViewBitmap(scrollView, picpath), picpath);

    }

    /**
     * 截取scrollview的屏幕
     **/
    public static Bitmap getNestedScrollViewBitmap(NestedScrollView scrollView, String picpath) {
        Bitmap bitmap;
        // 获取ScrollView实际高度
        scrollView.getChildAt(0).setBackgroundColor(Color.parseColor("#ffffff"));
        int height = scrollView.getHeight();
        int h = scrollView.getChildAt(0).getHeight();
        Log.d(TAG, "实际高度:" + h);
        Log.d(TAG, " 高度:" + height);
        if (h == 0 || height == 0) {
            return null;
        }
        // 创建对应大小的bitmap
        bitmap = Bitmap.createBitmap(scrollView.getWidth(), h, Bitmap.Config.ARGB_8888);
        final Canvas canvas = new Canvas(bitmap);
        scrollView.draw(canvas);
        // 测试输出
        FileOutputStream out = null;
        try {
            out = new FileOutputStream(picpath);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        try {
            if (null != out) {
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, out);
                out.flush();
                out.close();
            }
        } catch (IOException e) {
        }
        Log.d("TAG", "getScrollViewBitmap() returned: " + bitmap.getHeight());
        return bitmap;
    }


    // 截屏 ListView
    public static void shootListView(ListView listView, String picpath) {
        ScreenShotUtils.savePic(getListViewBitmap(listView, picpath), picpath);
    }

    /**
     * 截图listview
     **/
    public static Bitmap getListViewBitmap(ListView listView, String picpath) {
        int h = 0;
        Bitmap bitmap;
        // 获取listView实际高度
        for (int i = 0; i < listView.getChildCount(); i++) {
            h += listView.getChildAt(i).getHeight();
        }
        Log.d(TAG, "实际高度:" + h);
        Log.d(TAG, "list 高度:" + listView.getHeight());
        // 创建对应大小的bitmap
        bitmap = Bitmap.createBitmap(listView.getWidth(), h,
                Bitmap.Config.ARGB_8888);
        final Canvas canvas = new Canvas(bitmap);
        listView.draw(canvas);
        // 测试输出
        FileOutputStream out = null;
        try {
            out = new FileOutputStream(picpath);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        try {
            if (null != out) {
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, out);
                out.flush();
                out.close();
            }
        } catch (IOException e) {
        }
        return bitmap;
    }
    private static Bitmap getViewBitmapWithoutBottom(View v) {
        if (null == v) {
            return null;
        }
        v.setDrawingCacheEnabled(true);
        v.buildDrawingCache();
        if (Build.VERSION.SDK_INT >= 11) {
            v.measure(View.MeasureSpec.makeMeasureSpec(v.getWidth(), View.MeasureSpec.EXACTLY), View.MeasureSpec.makeMeasureSpec(v.getHeight(), View.MeasureSpec.EXACTLY));
            v.layout((int) v.getX(), (int) v.getY(), (int) v.getX() + v.getMeasuredWidth(), (int) v.getY() + v.getMeasuredHeight());
        } else {
            v.measure(View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED), View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
            v.layout(0, 0, v.getMeasuredWidth(), v.getMeasuredHeight());
        }
        Bitmap bp = Bitmap.createBitmap(v.getDrawingCache(), 0, 0, v.getMeasuredWidth(), v.getMeasuredHeight() - v.getPaddingBottom());
        v.setDrawingCacheEnabled(false);
        v.destroyDrawingCache();
        return bp;
    }

    public static Bitmap getViewBitmap(View v) {
        if (null == v) {
            return null;
        }
        v.setDrawingCacheEnabled(true);
        v.buildDrawingCache();
        if (Build.VERSION.SDK_INT >= 11) {
            v.measure(View.MeasureSpec.makeMeasureSpec(v.getWidth(), View.MeasureSpec.EXACTLY), View.MeasureSpec.makeMeasureSpec(v.getHeight(), View.MeasureSpec.EXACTLY));
            v.layout((int) v.getX(), (int) v.getY(), (int) v.getX() + v.getMeasuredWidth(), (int) v.getY() + v.getMeasuredHeight());
        } else {
            v.measure(View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED), View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
            v.layout(0, 0, v.getMeasuredWidth(), v.getMeasuredHeight());
        }
        Bitmap b = Bitmap.createBitmap(v.getDrawingCache(), 0, 0, v.getMeasuredWidth(), v.getMeasuredHeight());
        v.setDrawingCacheEnabled(false);
        v.destroyDrawingCache();
        return b;
    }

    /**
     * 获取 WebView 视图截图
     * @param context
     * @param view
     * @return
     */
    public static Bitmap getWebViewBitmap(Context context, WebView view) {
        if (null == view) return null;
        view.scrollTo(0, 0);
        view.buildDrawingCache(true);
        view.setDrawingCacheEnabled(true);
        view.setVerticalScrollBarEnabled(false);
        Bitmap b = getViewBitmapWithoutBottom(view);
        // 可见高度
        int vh = view.getHeight();
        // 容器内容实际高度
        int th = (int)(view.getContentHeight()*view.getScale());
        Bitmap temp = null;
        if (th > vh) {
            int w = getScreenWidth(context);
            int absVh = vh - view.getPaddingTop() - view.getPaddingBottom();
            do {
                int restHeight = th - vh;
                if (restHeight <= absVh) {
                    view.scrollBy(0, restHeight);
                    vh += restHeight;
                    temp = getViewBitmap(view);
                } else {
                    view.scrollBy(0, absVh);
                    vh += absVh;
                    temp = getViewBitmapWithoutBottom(view);
                }
                b = mergeBitmap(vh, w, temp, 0, view.getScrollY(), b, 0, 0);
            } while (vh < th);
        }
        // 回滚到顶部
        view.scrollTo(0, 0);
        view.setVerticalScrollBarEnabled(true);
        view.setDrawingCacheEnabled(false);
        view.destroyDrawingCache();
        return b;
    }


    /**
     * 拼接图片
     * @param newImageH
     * @param newImageW
     * @param background
     * @param backX
     * @param backY
     * @param foreground
     * @param foreX
     * @param foreY
     * @return
     */
    private static Bitmap mergeBitmap(int newImageH, int newImageW, Bitmap background, float backX, float backY, Bitmap foreground, float foreX, float foreY) {
        if (null == background || null == foreground) {
            return null;
        }
        Bitmap bitmap = Bitmap.createBitmap(newImageW, newImageH, Bitmap.Config.RGB_565);
        Canvas cv = new Canvas(bitmap);
        cv.drawBitmap(background, backX, backY, null);
        cv.drawBitmap(foreground, foreX, foreY, null);
        cv.save();
        cv.restore();
        return bitmap;
    }

    /**
     * get the width of screen
     */
    public static int getScreenWidth(Context ctx) {
        int w = 0;
        if (Build.VERSION.SDK_INT > 13) {
            Point p = new Point();
            ((WindowManager) ctx.getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay().getSize(p);
            w = p.x;
        } else {
            w = ((WindowManager) ctx.getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay().getWidth();
        }
        return w;
    }


    /**
     * webview截图
     * @param context
     * @param webView
     * @return
     */
    public static boolean getFullWebViewSnapshot(Context context,WebView webView) {
        //重新调用WebView的measure方法测量实际View的大小（将测量模式设置为UNSPECIFIED模式也就是需要多大就可以获得多大的空间）
        webView.measure(View.MeasureSpec.makeMeasureSpec(View.MeasureSpec.UNSPECIFIED, View.MeasureSpec.UNSPECIFIED),
                View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
        //调用layout方法设置布局（使用新测量的大小）
        webView.layout(0, 0, webView.getMeasuredWidth(), webView.getMeasuredHeight());
        //开启WebView的缓存(当开启这个开关后下次调用getDrawingCache()方法的时候会把view绘制到一个bitmap上)
        webView.setDrawingCacheEnabled(true);
        //强制绘制缓存（必须在setDrawingCacheEnabled(true)之后才能调用，否者需要手动调用destroyDrawingCache()清楚缓存）
        webView.buildDrawingCache();
        //根据测量结果创建一个大小一样的bitmap
        Bitmap picture = Bitmap.createBitmap(webView.getMeasuredWidth(),
                webView.getMeasuredHeight(), Bitmap.Config.ARGB_8888);
        //已picture为背景创建一个画布
        Canvas canvas = new Canvas(picture);  // 画布的宽高和 WebView 的网页保持一致
        Paint paint = new Paint();
        //设置画笔的定点位置，也就是左上角
        canvas.drawBitmap(picture, 0, webView.getMeasuredHeight(), paint);
        //将webview绘制在刚才创建的画板上
        webView.draw(canvas);
        //将bitmap保存到SD卡
        return  FileUtils.saveBitmap(context,picture,System.currentTimeMillis()+".JPEG");
    }


}

File: baselib/src/main/java/com/lishuaihua/baselib/util/ScreenShotUtils.kt
package com.lishuaihua.baselib.util

import android.app.Activity
import android.content.Context
import android.graphics.*
import android.os.Build
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.webkit.WebView
import android.widget.ListView
import androidx.core.widget.NestedScrollView
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException

object ScreenShotUtils {
    private const val TAG = "ScreenShotUtils"

    // 获取指定Activity的截屏，保存到png文件
    fun takeScreenShot(activity: Activity, navigationBarHeight: Int): Bitmap {
        // View是你需要截图的View
        val view = activity.window.decorView
        view.isDrawingCacheEnabled = true
        view.buildDrawingCache()
        val b1 = view.drawingCache

        // 获取状态栏高度
        val frame = Rect()
        activity.window.decorView.getWindowVisibleDisplayFrame(frame)
        val statusBarHeight = frame.top
        println(statusBarHeight)

        // 获取屏幕长和高
        val width = activity.windowManager.defaultDisplay.width
        val height = activity.windowManager.defaultDisplay
            .height
        // 去掉标题栏
        // Bitmap b = Bitmap.createBitmap(b1, 0, 25, 320, 455);
        val b = Bitmap.createBitmap(
            b1, 0, statusBarHeight + navigationBarHeight, width, height
                    - statusBarHeight - navigationBarHeight
        )
        view.destroyDrawingCache()
        return b
    }

    // 保存到sdcard
    fun savePic(b: Bitmap, strFileName: String?) {
        var fos: FileOutputStream? = null
        Log.d("TAG", "savePic() returned: ")
        try {
            fos = FileOutputStream(strFileName)
            if (null != fos) {
                b.compress(Bitmap.CompressFormat.PNG, 90, fos)
                Log.d("TAG", "savePic() returned:    " + b.height)
                fos.flush()
                fos.close()
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    // 截取当前屏幕
    fun shootScreenView(a: Activity, picpath: String?) {
        savePic(takeScreenShot(a, 0), picpath)
    }

    /**
     * 保存bitmap到SD卡
     * @param path 保存的名字
     * @param mBitmap 图片对像
     * return 生成压缩图片后的图片路径
     */
    fun saveMyBitmap(mBitmap: Bitmap?, path: String?): Boolean {
        val f = File(path)
        try {
            f.createNewFile()
        } catch (e: IOException) {
            println("在保存图片时出错：$e")
            return false
        }
        var fOut: FileOutputStream? = null
        fOut = try {
            FileOutputStream(f)
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
            return false
        }
        try {
            mBitmap!!.compress(Bitmap.CompressFormat.PNG, 100, fOut)
        } catch (e: Exception) {
            return false
        }
        try {
            fOut?.flush()
        } catch (e: IOException) {
            e.printStackTrace()
            return false
        }
        try {
            fOut?.close()
        } catch (e: IOException) {
            e.printStackTrace()
            return false
        }
        return true
    }

    // 程序入口 截取ScrollView
    fun shotNestedScrollView(scrollView: NestedScrollView, picpath: String?): Boolean {
        return saveMyBitmap(getNestedScrollViewBitmap(scrollView, picpath), picpath)
    }

    /**
     * 截取scrollview的屏幕
     */
    fun getNestedScrollViewBitmap(scrollView: NestedScrollView, picpath: String?): Bitmap? {
        val bitmap: Bitmap
        // 获取ScrollView实际高度
        scrollView.getChildAt(0).setBackgroundColor(Color.parseColor("#ffffff"))
        val height = scrollView.height
        val h = scrollView.getChildAt(0).height
        Log.d(TAG, "实际高度:$h")
        Log.d(TAG, " 高度:$height")
        if (h == 0 || height == 0) {
            return null
        }
        // 创建对应大小的bitmap
        bitmap = Bitmap.createBitmap(scrollView.width, h, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        scrollView.draw(canvas)
        // 测试输出
        var out: FileOutputStream? = null
        try {
            out = FileOutputStream(picpath)
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }
        try {
            if (null != out) {
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
                out.flush()
                out.close()
            }
        } catch (e: IOException) {
        }
        Log.d("TAG", "getScrollViewBitmap() returned: " + bitmap.height)
        return bitmap
    }

    // 截屏 ListView
    fun shootListView(listView: ListView, picpath: String?) {
        savePic(getListViewBitmap(listView, picpath), picpath)
    }

    /**
     * 截图listview
     */
    fun getListViewBitmap(listView: ListView, picpath: String?): Bitmap {
        var h = 0
        val bitmap: Bitmap
        // 获取listView实际高度
        for (i in 0 until listView.childCount) {
            h += listView.getChildAt(i).height
        }
        Log.d(TAG, "实际高度:$h")
        Log.d(TAG, "list 高度:" + listView.height)
        // 创建对应大小的bitmap
        bitmap = Bitmap.createBitmap(
            listView.width, h,
            Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(bitmap)
        listView.draw(canvas)
        // 测试输出
        var out: FileOutputStream? = null
        try {
            out = FileOutputStream(picpath)
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }
        try {
            if (null != out) {
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
                out.flush()
                out.close()
            }
        } catch (e: IOException) {
        }
        return bitmap
    }

    private fun getViewBitmapWithoutBottom(v: View?): Bitmap? {
        if (null == v) {
            return null
        }
        v.isDrawingCacheEnabled = true
        v.buildDrawingCache()
        if (Build.VERSION.SDK_INT >= 11) {
            v.measure(
                View.MeasureSpec.makeMeasureSpec(v.width, View.MeasureSpec.EXACTLY),
                View.MeasureSpec.makeMeasureSpec(v.height, View.MeasureSpec.EXACTLY)
            )
            v.layout(
                v.x.toInt(), v.y.toInt(), v.x.toInt() + v.measuredWidth, v.y
                    .toInt() + v.measuredHeight
            )
        } else {
            v.measure(
                View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
                View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
            )
            v.layout(0, 0, v.measuredWidth, v.measuredHeight)
        }
        val bp = Bitmap.createBitmap(
            v.drawingCache,
            0,
            0,
            v.measuredWidth,
            v.measuredHeight - v.paddingBottom
        )
        v.isDrawingCacheEnabled = false
        v.destroyDrawingCache()
        return bp
    }

    fun getViewBitmap(v: View?): Bitmap? {
        if (null == v) {
            return null
        }
        v.isDrawingCacheEnabled = true
        v.buildDrawingCache()
        if (Build.VERSION.SDK_INT >= 11) {
            v.measure(
                View.MeasureSpec.makeMeasureSpec(v.width, View.MeasureSpec.EXACTLY),
                View.MeasureSpec.makeMeasureSpec(v.height, View.MeasureSpec.EXACTLY)
            )
            v.layout(
                v.x.toInt(), v.y.toInt(), v.x.toInt() + v.measuredWidth, v.y
                    .toInt() + v.measuredHeight
            )
        } else {
            v.measure(
                View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
                View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
            )
            v.layout(0, 0, v.measuredWidth, v.measuredHeight)
        }
        val b = Bitmap.createBitmap(v.drawingCache, 0, 0, v.measuredWidth, v.measuredHeight)
        v.isDrawingCacheEnabled = false
        v.destroyDrawingCache()
        return b
    }

    /**
     * 获取 WebView 视图截图
     * @param context
     * @param view
     * @return
     */
    fun getWebViewBitmap(context: Context, view: WebView?): Bitmap? {
        if (null == view) return null
        view.scrollTo(0, 0)
        view.buildDrawingCache(true)
        view.isDrawingCacheEnabled = true
        view.isVerticalScrollBarEnabled = false
        var b = getViewBitmapWithoutBottom(view)
        // 可见高度
        var vh = view.height
        // 容器内容实际高度
        val th = (view.contentHeight * view.scale).toInt()
        var temp: Bitmap? = null
        if (th > vh) {
            val w = getScreenWidth(context)
            val absVh = vh - view.paddingTop - view.paddingBottom
            do {
                val restHeight = th - vh
                if (restHeight <= absVh) {
                    view.scrollBy(0, restHeight)
                    vh += restHeight
                    temp = getViewBitmap(view)
                } else {
                    view.scrollBy(0, absVh)
                    vh += absVh
                    temp = getViewBitmapWithoutBottom(view)
                }
                b = mergeBitmap(vh, w, temp, 0f, view.scrollY.toFloat(), b, 0f, 0f)
            } while (vh < th)
        }
        // 回滚到顶部
        view.scrollTo(0, 0)
        view.isVerticalScrollBarEnabled = true
        view.isDrawingCacheEnabled = false
        view.destroyDrawingCache()
        return b
    }

    /**
     * 拼接图片
     * @param newImageH
     * @param newImageW
     * @param background
     * @param backX
     * @param backY
     * @param foreground
     * @param foreX
     * @param foreY
     * @return
     */
    private fun mergeBitmap(
        newImageH: Int,
        newImageW: Int,
        background: Bitmap?,
        backX: Float,
        backY: Float,
        foreground: Bitmap?,
        foreX: Float,
        foreY: Float
    ): Bitmap? {
        if (null == background || null == foreground) {
            return null
        }
        val bitmap = Bitmap.createBitmap(newImageW, newImageH, Bitmap.Config.RGB_565)
        val cv = Canvas(bitmap)
        cv.drawBitmap(background, backX, backY, null)
        cv.drawBitmap(foreground, foreX, foreY, null)
        cv.save()
        cv.restore()
        return bitmap
    }

    /**
     * get the width of screen
     */
    fun getScreenWidth(ctx: Context): Int {
        var w = 0
        w = if (Build.VERSION.SDK_INT > 13) {
            val p = Point()
            (ctx.getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay.getSize(
                p
            )
            p.x
        } else {
            (ctx.getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay.width
        }
        return w
    }

    /**
     * webview截图
     * @param context
     * @param webView
     * @return
     */
    fun getFullWebViewSnapshot(context: Context, webView: WebView): Boolean {
        //重新调用WebView的measure方法测量实际View的大小（将测量模式设置为UNSPECIFIED模式也就是需要多大就可以获得多大的空间）
        webView.measure(
            View.MeasureSpec.makeMeasureSpec(
                View.MeasureSpec.UNSPECIFIED,
                View.MeasureSpec.UNSPECIFIED
            ),
            View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        )
        //调用layout方法设置布局（使用新测量的大小）
        webView.layout(0, 0, webView.measuredWidth, webView.measuredHeight)
        //开启WebView的缓存(当开启这个开关后下次调用getDrawingCache()方法的时候会把view绘制到一个bitmap上)
        webView.isDrawingCacheEnabled = true
        //强制绘制缓存（必须在setDrawingCacheEnabled(true)之后才能调用，否者需要手动调用destroyDrawingCache()清楚缓存）
        webView.buildDrawingCache()
        //根据测量结果创建一个大小一样的bitmap
        val picture = Bitmap.createBitmap(
            webView.measuredWidth,
            webView.measuredHeight, Bitmap.Config.ARGB_8888
        )
        //已picture为背景创建一个画布
        val canvas = Canvas(picture) // 画布的宽高和 WebView 的网页保持一致
        val paint = Paint()
        //设置画笔的定点位置，也就是左上角
        canvas.drawBitmap(picture, 0f, webView.measuredHeight.toFloat(), paint)
        //将webview绘制在刚才创建的画板上
        webView.draw(canvas)
        //将bitmap保存到SD卡
        return FileUtils.saveBitmap(
            context,
            picture,
            System.currentTimeMillis().toString() + ".JPEG"
        )
    }
}